### PR TITLE
[SharkInference-SharkRuntime] Adds capability to mmap vmfbs

### PR DESCRIPTION
**Pull request by Abhishek-Varma**
-- This commit is based on [VmModule.mmap() API](https://github.com/openxla/iree/pull/14124).
-- It thereby adds capability to mmap vmfbs in SHARK.

Signed-off-by: Abhishek Varma <abhishek@nod-labs.com>

### CODE REVIEW

**Comment by Abhishek-Varma** [comment]
So, `tempfile.NamedTemporaryFile` needs to have `delete=False` set for us to get this to work on Windows.

Also, on Windows I see an API related issue in `VmModule.mmap` at this [line](https://github.com/openxla/iree/blob/ff086159340967c82834a6a7b55e4f9ae9ef285c/runtime/bindings/python/vm.cc#L158). The error being :-
```
AttributeError: module 'mmap' has no attribute 'MAP_SHARED'
```

**Comment by powderluv** [comment]
This has landed again. https://github.com/openxla/iree/pull/14153

File: shark/iree_utils/compile_utils.py [hunk 1]
```diff
@@ -316,26 +317,102 @@ def get_iree_module(flatbuffer_blob, device, device_idx=None):
...
+    #   OR 2. We are compiling on the fly, therefore we have the flatbuffer blob to play with.
+    #         (This would arise if we're invoking `compile` from a SharkInference obj)
+    if (
+        isinstance(flatbuffer_blob_or_path, str)
+        and ".vmfb" in flatbuffer_blob_or_path
+    ):
+        mmaped_vmfb = ireert.VmModule.mmap(instance, flatbuffer_blob_or_path)
+        context = ireert.VmContext(instance, modules=[hal_module, mmaped_vmfb])
+    else:
+        tmpf = tempfile.NamedTemporaryFile(delete=False)
```
Comment (powderluv) [inline shark/iree_utils/compile_utils.py:357]: this should be automatic now.

**Comment by powderluv** [comment]
Lets land this tomorrow so we can have mmap support

**Comment by Abhishek-Varma** [comment]
Oh! I missed this - sure, I'll take a look at the upstream changes - accordingly make changes to this PR, test and mark this ready for review.

**Comment by Abhishek-Varma** [comment]
So, I'm trying to see how to unlink the mapped file.
Because the callback leads to an error because by the time it is triggered the temporary file's name/data is lost.
Will test on Windows too post that - I see my setup was deleted.

Worst case - I'll keep the mmap loading confined only to those cases where we indeed have the vmfb generated and saved. Therefore getting rid of the temporary file's need.

Have made changes to `from_flatbuffer` API to use `from_buffer` API + the warning being set off.

Will update here on my progress.

@powderluv

**Comment by powderluv** [comment]
Your in python 3.10. using 3.11 will fix it

**Review by monorimet — APPROVED** [review]
LGTM! Thanks for this.

**Pull request merged**